src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java

```diff
@@ -61,8 +55,7 @@ static int getNumber(final String methodName, final Class<?>[] argTypes, final C
     }
     final Class<?>[] t1 = o1.getParameterTypes();
     final Class<?>[] t2 = o2.getParameterTypes();
-    // both null
-    if (t1 == t2) {
+    if ((t1 == null) && (t2 == null)) {
```

Comment:
This complete method/lambda can be replaced with
```java
Comparator.comparing(Method::getName)
  .thenComparing(Method::getParameterTypes,
    Comparator.comparing(o -> o == null)
    .thenComparingInt(a -> a.length)
    .thenComparing((o1, o2) -> {/*insert for-loop array comparison*/});

Reply:
It seems the last thenComparing bit is not required.

Reply:
Or that the last thenComparing bit needs to be
`.thenComparing(Function.identity())`

Reply:
Trying this out:
```
  private static final Comparator<Method> methodComparator = Comparator.comparing(Method::getName)
      .thenComparing(Method::getParameterTypes, typeName ->
          Comparator.comparing(o -> o == null)
              .thenComparingInt(a -> a.length)
              .thenComparing((o1, o2) -> {/*insert for-loop array comparison*/});

```
I get a compilation error on `a.length`

Reply:
Fixed the bad arrow: length is wrong type.
```
  private static final Comparator<Method> methodComparator = Comparator.comparing(Method::getName)
      .thenComparing(Method::getParameterTypes,
          Comparator.comparing(o -> o == null)
              .thenComparingInt(a -> a.length)
              .thenComparing(Function.identity())
              .thenComparing((o1, o2) -> {/*insert for-loop array comparison*/});
```

Reply:
Perhaps using simply `thenComparing` instead of `thenComparingInt`?

Reply:
Hmm,  gave a few things a try, no luck. Can pick back up later. One caution, we need to be careful about current null handling, I wonder if comparator could change that on us?

Reply:
The thenComparing statements are short-circuiting AFAIK.
Couldn't find this explicitly mentioned anywhere, but at least the OpenJDK implementation is.

Reply:
I wonder if
```java
Comparator.comparing(Method::getName)
  .thenComparing(Method::getParameterTypes);
``` 
would be sufficient for the sorting order.

Reply:
It also seems `Method#getParameterTypes` doesn't return 0

Reply:
It seems we need to explicitly specify the type argument:
```java
Comparator.comparing(Method::getName)
  .thenComparing(Method::getParameterTypes,
    Comparator.comparingInt((Class<?>[] a) -> a.length)
    .thenComparing((o1, o2) -> {
      for (int i = 0; i < o1.length; i++) {
        final int compareValue = o1[i].getName().compareTo(o2[i].getName());
        if (compareValue != 0) {
          return compareValue;
        }
        return 0;
      }));
```
This snippet works for me

Reply:
Hmm, I'm not having much luck! :disappointed: 

```

  private void sortMethods(Method[] methods) {
    Arrays.sort(methods, Comparator.comparing(Method::getName)
        .thenComparing(Method::getParameterTypes,
            Comparator.comparingInt((Class<?>[] a) -> a.length)
                .thenComparing((o1, o2) -> {
                  for (int i = 0; i < o1.length; i++) {
                    final int compareValue = o1[i].getName().compareTo(o2[i].getName());
                    if (compareValue != 0) {
                      return compareValue;
                    }
                    return 0;
                  }
                  })));
```
![screenshot from 2018-01-02 21-52-45](https://user-images.githubusercontent.com/12397753/34510273-5b30eee0-f007-11e7-9bb1-a7439013fcda.png)

I'm pretty tempted to just leave this method alone to move on for now, it's okay as is...  : |

Reply:
No problem, I'll submit a follow-up PR.

Reply:
Unless, of course, @ssoloff figures out a better way that doesn't require that much explicit typisation like the last time when I insited that this type specification was absolutely required 😬 ^^